ktor-server/ktor-server-cio/jvmAndNix/src/io/ktor/server/cio/CIOApplicationEngine.kt

```diff
@@ -113,6 +117,51 @@ public class CIOApplicationEngine(
         }
     }
 
+    private suspend fun addHandlerForExpectedHeader(output: ByteWriteChannel, call: CIOApplicationCall) {
+        val continueResponse = "HTTP/1.1 100 Continue\r\n"
+        val expectHeaderValue = "100-continue"
+
+        val expectHeader = call.request.headers[HttpHeaders.Expect]?.lowercase()
+        var closeConnection = true
+
+        val expectedHeaderPhase = PipelinePhase("ExpectedHeaderPhase")
+        call.request.pipeline.insertPhaseBefore(ApplicationReceivePipeline.Before, expectedHeaderPhase)
+        call.request.pipeline.intercept(expectedHeaderPhase) {
+            val request = call.request
+            val version = HttpProtocolVersion.parse(request.httpVersion)
+            val hasBody = hasBody(request)
+
+            if (expectHeader == null || version == HttpProtocolVersion.HTTP_1_0 || !hasBody) {
+                return@intercept
+            }
+
+            if (expectHeader != expectHeaderValue) {
+                call.respond(HttpStatusCode.ExpectationFailed)
+            } else {
+                output.apply {
+                    output.writeStringUtf8(continueResponse)
+                    output.flush()
+                }
+                closeConnection = false
+            }
+        }
+
+        call.response.pipeline.intercept(ApplicationSendPipeline.Transform) {
+            if (expectHeader == null) {
+                return@intercept
+            }
+
+            val connection = if (closeConnection) "close" else "keep-alive"
+            call.response.headers.append(HttpHeaders.Connection, connection)
+        }
+    }
+
+    private fun hasBody(request: CIOApplicationRequest): Boolean {
```

Comment:
@e5l Can you check please if this method is correct? Are there any other cases that we miss here?

Reply:
```
The presence of a message-body in a request is signaled by the
inclusion of a Content-Length or Transfer-Encoding header field in
the request's message-headers.
```
https://www.rfc-editor.org/rfc/rfc2616.html#section-4.3